pkg/suggestion/v1alpha3/goptuna/sample.go

```diff
@@ -0,0 +1,146 @@
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: strconv.FormatFloat(p, 'f', -1, 64),
+			})
+		case goptuna.DiscreteUniformDistribution:
+			p, err := trial.SuggestDiscreteFloat(name, distribution.Low, distribution.High, distribution.Q)
+			if err != nil {
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: strconv.FormatFloat(p, 'f', -1, 64),
+			})
+		case goptuna.IntUniformDistribution:
+			p, err := trial.SuggestInt(name, distribution.Low, distribution.High)
+			if err != nil {
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: strconv.Itoa(p),
+			})
+		case goptuna.StepIntUniformDistribution:
+			p, err := trial.SuggestStepInt(name, distribution.Low, distribution.High, distribution.Step)
+			if err != nil {
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: strconv.Itoa(p),
+			})
+		case goptuna.CategoricalDistribution:
+			p, err := trial.SuggestCategorical(name, distribution.Choices)
+			if err != nil {
+				klog.Errorf("Failed to get suggested param: %s", err)
+				return nil, err
+			}
+			assignments = append(assignments, &api_v1_alpha3.ParameterAssignment{
+				Name:  name,
+				Value: p,
+			})
+		}
+	}
+
+	if t, err := study.Storage.GetTrial(trial.ID); err == nil {
+		klog.Infof("Success to sample new trial: trialID=%d, params=%v", t.ID, t.Params)
+	}
+	return assignments, nil
+}
+
+func isSameTrialParam(ktrial, gtrial goptuna.FrozenTrial) bool {
+	// Compare trial parameters by "internal representation".
+	// In the internal representation, all parameters are represented by `float64` to store the storage
+	// (because Goptuna supports not only in-memory but also RDB storage backend).
+	// To represent categorical parameters, Goptuna holds an index of the list in the database.
+	//
+	// SearchSpace: map[string]interface{}{"x1": Uniform{Min: -10, Max: 10}, "x2": Categorical{Choices: []string{"param-1", "param-2"}}}
+	// External representation: map[string]interface{}{"x1": 5.5, "x2": "param-2"}
+	// Internal representation: map[string]float64{"x1": 5.5, "x2": 1.0}
+	for name := range gtrial.InternalParams {
+		gtrialParamValue := gtrial.InternalParams[name]
+		ktrialParamValue, ok := ktrial.InternalParams[name]
+		if !ok {
+			// must not reach here
+			klog.Errorf("Detect inconsistent internal parameters: %v and %v",
+				ktrial.InternalParams, gtrial.InternalParams)
+			return false
+		}
+		if gtrialParamValue != ktrialParamValue {
+			return false
+		}
+	}
+	return true
+}
```

Comment:
Sorry, this logic seems to be broken.

In Goptuna, the internal representation is not necessarily reproducible from external representation because of the following reason:
https://github.com/optuna/optuna/issues/925

So we need to compare external representations because `ktrial` is created from Katib parameter assignments which store external representation. I'll fix this soon.

Reply:
I think it is enough to use `reflect.DeepEqual(gtrial.Params, ktrial.Params`.

Reply:
I fixed this at 8c21693 and tested on the following parameter settings:

```
# https://github.com/c-bata/katib-goptuna-example/blob/master/experiment-step.yaml
  parameters:
    - name: x1
      parameterType: int
      feasibleSpace:
        min: "-15"
        max: "15"
        step: "2"
    - name: x2
      parameterType: double
      feasibleSpace:
        min: "-10"
        max: "10"
        step: 0.5
```

It works fine!

```
>>> logs -f example-step3-cmaes-54b4886448-cgkdm
I0414 06:22:33.668875       1 main.go:36] Start Goptuna suggestion service: 0.0.0.0:6789
I0414 06:22:47.179106       1 service.go:67] Success to sample new trial: trialID=0, assignments=[name:"x1" value:"3"  name:"x2" value:"1.5" ]
I0414 06:22:47.182047       1 service.go:67] Success to sample new trial: trialID=1, assignments=[name:"x1" value:"5"  name:"x2" value:"-1" ]
I0414 06:22:53.962858       1 service.go:100] Update trial mapping : trialName=example-step3-lvmdgldl -> trialID=1
I0414 06:22:53.962882       1 service.go:100] Update trial mapping : trialName=example-step3-458v8s2c -> trialID=0
I0414 06:22:53.962888       1 service.go:130] Detect changes of Trial (trialName=example-step3-458v8s2c, trialID=0) : State Complete, Evaluation 46.250000
I0414 06:22:53.962988       1 service.go:67] Success to sample new trial: trialID=2, assignments=[name:"x2" value:"0.5"  name:"x1" value:"-3" ]
I0414 06:22:56.158324       1 service.go:100] Update trial mapping : trialName=example-step3-mrxdqpq9 -> trialID=2
I0414 06:22:56.158345       1 service.go:130] Detect changes of Trial (trialName=example-step3-lvmdgldl, trialID=1) : State Complete, Evaluation 16.000000
I0414 06:22:56.158473       1 service.go:67] Success to sample new trial: trialID=3, assignments=[name:"x1" value:"-7"  name:"x2" value:"1" ]
I0414 06:22:59.670946       1 service.go:130] Detect changes of Trial (trialName=example-step3-mrxdqpq9, trialID=2) : State Complete, Evaluation 94.250000
I0414 06:22:59.670984       1 service.go:100] Update trial mapping : trialName=example-step3-z9p7rklk -> trialID=3
I0414 06:22:59.671069       1 service.go:67] Success to sample new trial: trialID=4, assignments=[name:"x1" value:"-1"  name:"x2" value:"-1" ]
```

PTAL.